replica/table.cc

```diff
@@ -2428,6 +2458,51 @@ future<> tablet_storage_group_manager::handle_tablet_split_completion(const loca
     return stop_fut;
 }
 
+future<> tablet_storage_group_manager::handle_tablet_merge_completion(const locator::tablet_map& old_tmap, const locator::tablet_map& new_tmap) {
+    auto table_id = schema()->id();
+    size_t old_tablet_count = old_tmap.tablet_count();
+    size_t new_tablet_count = new_tmap.tablet_count();
+    storage_group_map new_storage_groups;
+
+    unsigned log2_reduce_factor = log2ceil(old_tablet_count / new_tablet_count);
+    unsigned merge_size = 1 << log2_reduce_factor;
+
+    if (merge_size != 2) {
+        throw std::runtime_error(format("Tablet count was not reduced by a factor of 2 (old: {}, new {}) for table {}",
+                                 old_tablet_count, new_tablet_count, table_id));
+    }
+
+    for (auto& [id, sg] : _storage_groups) {
+        // Pick first (even) tablet of each sibling pair.
+        if (id % merge_size != 0) {
+            continue;
+        }
+        auto new_tid = id >> log2_reduce_factor;
+
+        auto new_cg = make_lw_shared<compaction_group>(_t, new_tid, new_tmap.get_token_range(locator::tablet_id(new_tid)));
+        auto new_sg = make_lw_shared<storage_group>(std::move(new_cg));
+
+        for (unsigned i = 0; i < merge_size; i++) {
+            auto group_id = id + i;
+
+            auto it = _storage_groups.find(group_id);
+            if (it == _storage_groups.end()) {
+                throw std::runtime_error(format("Unable to find sibling tablet of id for table {}", group_id, table_id));
+            }
+            auto& sg = it->second;
+            sg->for_each_compaction_group([&new_sg, new_tid] (const compaction_group_ptr& cg) {
+                cg->update_id(new_tid);
+                new_sg->add_merging_group(cg);
+            });
+            // FIXME: we MUST schedule a background action to flush memtable and move sstables of merging groups into the main one.
```

Comment:
I thought about this, but it is complicated since memtable_list has a lambda that references its underlying compaction_group, and ongoing flush takes compaction group as a reference (see table::make_memtable_list()). So if we just move the list, we end up with a dangling pointer. We cannot afford to wait for flush here either, since the changes in replica state must be applied atomically. This why I need merging groups. The groups will be eventually merged into main group of new tablet. Maybe we can optimize it later with some work in table::seal_active_memtable() where it works with a reference to memtable list, and grabs the compaction group only when needed (e.g. when adding the sstable produced to the set).